### PR TITLE
Migrate to EdgeSpec

### PR DESCRIPTION
**Pull request by codetheweb**
I'll bump to Node 20 in a separate PR.

### CODE REVIEW

File: src/api/things.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+const jsonResponse = z.object({
+  thing: m.thing.optional(),
+  things: z.array(m.thing).optional(),
+})
+
+export default withRouteSpec({
+  auth: "api_key",
+  methods: ["GET", "POST"],
+  jsonBody: jsonBody.optional(),
```
Comment (mxsdev) [inline src/api/things.ts:16]: Shouldn't it be possible to not have this be optional w/ `shouldValidateGetRequestBody: false`?

Comment (mxsdev) [reply src/api/things.ts:16]: Or I guess we are removing `shouldValidateGetRequestBody` with method discrimination?

File: src/lib/make-request.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { loadBundle } from "edgespec"
+import type { Middleware } from "edgespec/middleware"
+
+import { createDatabase, type Database } from "./database/index.js"
+
+export const createMakeRequest = async (database?: Database) => {
+  const bundle = await loadBundle("./bundled-routes.js")
```
Comment (mxsdev) [inline src/lib/make-request.ts:7]: nice...so this avoid the chicken-and-egg problem of having to uild before running the server fixture?

**Review by mxsdev — APPROVED** [review]

File: src/lib/make-request.ts (hunk 2 above)
Comment (codetheweb) [reply src/lib/make-request.ts:7]: yep, we can just import at runtime 👍 
although I guess this doesn't work for wintercg envs so maybe we should change this

File: .eslintrc.json [hunk 3]
```diff
@@ -47,14 +47,7 @@
...
       },
-      "overrides": [
-        {
-          "files": ["./next-env.d.ts"],
-          "rules": {
-            "@typescript-eslint/triple-slash-reference": "off"
-          }
-        }
-      ],
+      "overrides": [],
```
Comment (razor-x) [inline .eslintrc.json:50]:
```suggestion
```

File: .gitignore
Comment (razor-x) [inline .gitignore:1]: Use git add -p

File: .github/workflows/check.yml [hunk 4]
```diff
@@ -101,7 +101,11 @@ jobs:
           write-mode: overwrite
           path: index.js
           contents: |
-            import '@seamapi/fake-template'
+            import {createFake} from '@seamapi/fake-template'
+            const fake = await createFake()
+
+            const response = await fake.makeRequest(new Request("https://example.com/health"))
+            console.log(response)
```
Comment (razor-x) [inline .github/workflows/check.yml:108]: Assert something about response?

File: src/lib/fake.ts [hunk 5]
```diff
@@ -2,45 +2,36 @@ import {
...
-
-  async stopServer(): Promise<void> {
-    this.server?.close()
-  }
-
-  get serverUrl(): string | null | undefined {
-    return this.server?.serverUrl
-  }
-
-  async seed(): Promise<Seed> {
```
Comment (razor-x) [inline src/lib/fake.ts:42]: Add this back

File: examples/database-sample.ts [hunk 6]
```diff
@@ -24,7 +24,7 @@ export const builder: Builder = {
 
 export const handler: Handler<Options> = async ({ outfile, logger }) => {
   const fake = await createFake()
-  await fake.seed()
+  seedDatabase(fake.database)
```
Comment (razor-x) [inline examples/database-sample.ts:27]: Revert this

File: src/pages/index.tsx
Comment (razor-x) [inline src/pages/index.tsx:1]: Restore support for this.

File: src/route-types.ts [hunk 7]
```diff
@@ -1,50 +1,30 @@
...
-            thingId: string
-            type: "superthing" | "lamething"
-            status: "online" | "offline"
-          }[]
-        | undefined
-    }
-  }
-}
-
-export type RouteResponse<Path extends keyof Routes> =
```
Comment (razor-x) [inline src/route-types.ts:43]: Can edgespec still generate these?

Comment (codetheweb) [reply src/route-types.ts:43]: https://github.com/seamapi/edgespec/pull/139

File: src/pages/index.tsx
Comment (codetheweb) [reply src/pages/index.tsx:1]: d36f4fd

File: src/api/things.ts (hunk 1 above)
Comment (codetheweb) [reply src/api/things.ts:16]: personally I think we should remove `shouldValidateGetRequestBody` and prefer the `.optional()` style
in the HTTP spec it's invalid to provide a body for GET so I think we can probably just ignore it

File: tsconfig.json [hunk 8]
```diff
@@ -1,16 +1,20 @@
...
     "strict": true,
     "esModuleInterop": true,
     "moduleResolution": "nodenext",
+    "module": "NodeNext",
     "allowJs": false,
     "target": "es2021",
     "declaration": true,
     "sourceMap": true,
     "newLine": "lf",
+    "jsx": "react",
```
Comment (razor-x) [inline tsconfig.json:17]: Ideally we would use this, is it possible?
```suggestion
    "jsx": "react-jsx",
```

File: tsconfig.json [hunk 9]
```diff
@@ -29,17 +33,16 @@
...
     "src/**/*",
-    "next-env.d.ts",
-    ".next/types/**/*.ts",
+    "edgespec.config.ts",
     "test/**/*",
     "examples/**/*",
     "tsup.config.ts"
-  ]
+  ],
+  "exclude": ["dist"]
```
Comment (razor-x) [inline tsconfig.json:47]: Why is this needed now?

File: package.json [hunk 10]
```diff
@@ -28,21 +28,17 @@
     "!**/*.test.ts"
   ],
   "scripts": {
-    "start": "next dev",
-    "build": "npm run build:ts",
+    "dev": "edgespec dev",
+    "build": "npm run build:ts && edgespec bundle -o dist/bundled-routes.js",
```
Comment (razor-x) [inline package.json:32]: Reminder to make issue to update the scripts to not use &&

File: package.json [hunk 11]
```diff
@@ -28,21 +28,17 @@
...
     "build:ts": "tsup",
     "prebuild:ts": "del dist",
     "postbuild:ts": "tsup --format cjs src/index.ts",
-    "build:nsm": "nsm build",
-    "build:route-types": "nextlove generate-route-types --packageDir . --outputFile ./src/route-types.ts --apiPrefix / --pathGlob '/src/pages/api/**/*.ts'",
+    "build:route-types": "edgespec codegen route-types -o src/route-types.ts",
     "postbuild:route-types": "prettier --write src/route-types.ts",
-    "build:openapi": "nextlove generate-openapi --packageDir . --outputFile ./public/openapi.json --apiPrefix / --pathGlob '/src/pages/api/**/*.ts'",
-    "prebuild:openapi": "mkdirp public",
+    "build:openapi": "edgespec codegen openapi -o public/openapi.json",
```
Comment (razor-x) [inline package.json:41]: I wonder if we should even generate this, vs. can we serve it from the fake server? 

See, e.g., https://github.com/seamapi/javascript-http/blob/main/test/seam/connect/openapi.test.ts#L11

File: package.json [hunk 12]
```diff
@@ -86,16 +83,15 @@
...
-    "nextlove": "^2.6.0",
     "prettier": "^3.0.3",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
     "tsup": "^8.0.1",
     "tsx": "^4.6.2",
-    "typed-axios-instance": "^3.1.0",
-    "typescript": "~5.1.0",
+    "typed-axios-instance": "^3.3.1",
+    "typescript": "~5.3.3",
```
Comment (razor-x) [inline package.json:94]: Can go back to this now that upgrading is safe again
```suggestion
    "typescript": "^5.3.3",
```

File: src/routes/index.tsx [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+import { withRouteSpec } from "lib/middleware/index.ts"
+import { renderJsxToResponse } from "lib/render-jsx-to-response.ts"
+
+export default withRouteSpec({
+  auth: "none",
+  methods: ["GET"],
+  middlewares: [],
+})(async (_req, _ctx) => {
+  return await renderJsxToResponse(<h1>Hello world!</h1>)
```
Comment (razor-x) [inline src/routes/index.tsx:11]: WTB middleware for
```suggestion
  return await ctx.jsx(<h1>Hello world!</h1>)
```

File: src/lib/render-jsx-to-response.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import type React from "react"
+import reactDom from "react-dom/server"
+
+export const renderJsxToResponse = async (element: React.ReactNode) => {
```
Comment (razor-x) [inline src/lib/render-jsx-to-response.ts:4]:
```suggestion
import type { ReactNode } from "react"
import reactDom from "react-dom/server"

export const renderJsxToResponse = async (element: ReactNode) => {
```

File: test/fixtures/get-test-server.ts [hunk 15]
```diff
@@ -1,54 +1,46 @@
...
+  }) as TypedAxios<Routes>
+
+  axios.defaults.headers.common["authorization"] = `Bearer ${seed.api_key}`
 
   return {
-    ...fixture,
-    // @ts-expect-error Current version of axios has upstream type issue.
-    get: fixture.axios.get.bind(fixture.axios),
+    serverURL,
+    axios,
```
Comment (razor-x) [inline test/fixtures/get-test-server.ts:42]: Should we move to this now?
```suggestion
    client,
```

File: src/routes/things.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  methods: ["GET", "POST"],
+  jsonBody: jsonBody.optional(),
+  jsonResponse,
+})(async (req, ctx) => {
+  if (req.method === "GET") {
+    return ctx.json({ things: ctx.db.things })
+  }
+
+  if (req.method === "POST") {
+    // todo: enable method discrimination upstream (https://github.com/seamapi/edgespec/issues/136)
```
Comment (razor-x) [inline src/routes/things.ts:24]:
```suggestion
    // UPSTREAM: edgespec need a pattern for method discrimination.
    // https://github.com/seamapi/edgespec/issues/136
```

File: src/lib/fake.ts [hunk 17]
```diff
@@ -6,43 +6,40 @@ import {
...
-  get serverUrl(): string | null | undefined {
-    return this.server?.serverUrl
+    this.makeRequestWithDatabase = createMakeRequest(database)
   }
 
   async seed(): Promise<Seed> {
     return seedDatabase(this.#database)
   }
 
+  async makeRequest(request: Request): Promise<Response> {
```
Comment (razor-x) [inline src/lib/fake.ts:37]: Where is Response defined? I don't see it imported anywhere.

Comment (razor-x) [inline src/lib/fake.ts:37]: Would be neat if this was just `fetch` and implemented the fetch API 🤔

**Review by razor-x — APPROVED** [review]
I don't think any comments are blocked, but there are few we might want to get in on this PR. Some may be future upgrades.

File: tsconfig.json [hunk 18]
```diff
@@ -1,16 +1,20 @@
 {
   "$schema": "https://json.schemastore.org/tsconfig",
-  "extends": "@tsconfig/next/tsconfig.json",
   "compilerOptions": {
+    "lib": ["dom", "dom.iterable", "esnext"],
```
Comment (razor-x) [inline tsconfig.json:4]: Do we want this / need this anymore? I think `dom` and `dom.iterable` don't work anymore since we don`t use next react.

```suggestion
```

**Review by razor-x — COMMENTED** [review]
I don't think any comments are blocked, but there are few we might want to get in on this PR. Some may be future upgrades.

File: tsconfig.json (hunk 8 above)
Comment (codetheweb) [reply tsconfig.json]: yep 👍 
33243f8

File: tsconfig.json (hunk 9 above)
Comment (codetheweb) [reply tsconfig.json]: not needed

File: src/lib/fake.ts (hunk 17 above)
Comment (codetheweb) [reply src/lib/fake.ts:37]: it's from @types/node, same as Request

Comment (codetheweb) [reply src/lib/fake.ts:37]: the `fetch` API is *basically* `Request`

but yeah I wouldn't be opposed to `makeRequest() -> fetch()`

File: tsconfig.json (hunk 18 above)
Comment (codetheweb) [reply tsconfig.json:4]: they still work, they're provided by TS

needed for JSX rendering:

Screenshot 2024-03-07 at 17 30 11

File: test/fixtures/get-test-server.ts (hunk 15 above)
Comment (codetheweb) [reply test/fixtures/get-test-server.ts:42]: yep works for me

File: package.json (hunk 11 above)
Comment (codetheweb) [reply package.json:41]: interesting... I'd be ok generating it on request in the dev server but including that logic in the bundle may bloat it by quite a bit
not sure how heavy the openapi builder package is

File: src/routes/index.tsx (hunk 13 above)
Comment (codetheweb) [reply src/routes/index.tsx:9]: can make an issue on edgespec for it, happy to add as long as it doesn't require all consumers to install & bundle React

Comment (razor-x) [reply src/routes/index.tsx:9]: Could be the first edgespec plugin. I think `@edgespec/jsx` or  `@edgespec/react` would be need

File: src/lib/make-request.ts (hunk 2 above)
Comment (codetheweb) [reply src/lib/make-request.ts]: turned out a lot simpler than I was thinking

2e23274

File: src/lib/make-request.ts [hunk 19]
```diff
@@ -1,11 +1,10 @@
-import { loadBundle } from "edgespec"
 import type { Middleware } from "edgespec/middleware"
 
+// @ts-expect-error (otherwise this will throw during `edgespec codegen`)
+import bundle from "../../dist/bundled-routes.js"
```
Comment (razor-x) [inline src/lib/make-request.ts:4]: Is there an upstream issue we can link to in edgespec?

Comment (codetheweb) [reply src/lib/make-request.ts:4]: this isn't really an edgespec issue, `npx edgespec codegen route-types` uses the TypeScript AST/checker so your code has to type check successfully before you can generate types

Comment (codetheweb) [reply src/lib/make-request.ts:4]: we could change the ordering so that instead `npx edgespec bundle` runs before `npx edgespec codegen`, but then `npm run codegen` will fail if you run it outside of `npm run build`.

Comment (razor-x) [reply src/lib/make-request.ts:4]: For `build:*` scripts I see them as a unit, e.g., `npm run build` will work but individual steps may not work if you run them manually out of order.

Comment (codetheweb) [reply src/lib/make-request.ts:4]: personally I would rather keep the ability to run `npm run codegen` outside of build; I commonly use it to update route types since the route types are used in tests (typed-axios)
a single @ts-expect-error seems like a minor tradeoff for that
